enterprise/backend/test/metabase_enterprise/caching/strategies_test.clj

```diff
@@ -0,0 +1,257 @@
+                      (is (=? (mkres (t 0))
+                              (-> (qp/process-query q) (dissoc :data)))))
+                    (testing "No cache after expiration"
+                      (mt/with-clock (t 61)
+                        (is (=? (mkres nil)
+                                (-> (qp/process-query q) (dissoc :data))))))))))
+
+            (testing "strategy = schedule"
+              (mt/with-model-cleanup [[:model/QueryCache :updated_at]]
+                (mt/with-clock (t 0)
+                  (is (pos? (#'task.caching/refresh-schedule-configs)))
+                  (let [q (#'qp.card/query-for-card card3 {} {} {} {})]
+                    (is (=? {:type :schedule}
+                            (:cache-strategy q)))
+                    (is (=? (mkres nil)
+                            (-> (qp/process-query q) (dissoc :data))))
+                    (testing "There is cache on second call"
+                      (is (=? (mkres (t 0))
+                              (-> (qp/process-query q) (dissoc :data)))))))
+                (testing "No cache after job ran again"
+                  (mt/with-clock (t 121)
+                    (is (pos? (#'task.caching/refresh-schedule-configs)))
+                    (let [q (#'qp.card/query-for-card card3 {} {} {} {})]
+                      (is (=? (mkres nil)
+                              (-> (qp/process-query q) (dissoc :data)))))))))
+
+            (testing "strategy = query"
+              (mt/with-model-cleanup [[:model/QueryCache :updated_at]]
+                (mt/with-clock (t 0)
+                  (is (pos? (#'task.caching/refresh-query-configs)))
+                  (let [q (#'qp.card/query-for-card card4 {} {} {} {})]
+                    (is (=? {:type :query}
+                            (:cache-strategy q)))
+                    (is (=? (mkres nil)
+                            (-> (qp/process-query q) (dissoc :data))))
+                    (testing "There is cache on second call"
+                      (is (= 1 (count (t2/select :model/QueryCache))))
+                      (is (=? (mkres (t 0))
+                              (-> (qp/process-query q) (dissoc :data)))))))
+
+                (mt/with-clock (t 121)
+                  (is (pos? (#'task.caching/refresh-query-configs)))
+                  (testing "Nothing to run, because it's already scheduled for later"
+                    (is (zero? (#'task.caching/refresh-query-configs))))
+                  (testing "Job ran again, but state has not changed, so there's still cache"
+                    (let [q (#'qp.card/query-for-card card4 {} {} {} {})]
+                      (is (=? (mkres (t 0))
+                              (-> (qp/process-query q) (dissoc :data)))))))
+
+                (mt/with-clock (t 242)
+                  (t2/update! :model/CacheConfig {:id (:id c4)} {:config (assoc (:config c4)
+                                                                                :field_id (mt/id :table :value))})
```

Comment:
It feels a bit odd to use changes to config parameters to reset, why not change the `:marker` in the `:state` instead, to focus on the primary mechanism?

Reply:
Makes sense, but maybe we want both then?